File: problems/376-wiggle-subsequence/BottomupDP-v2.cpp
/**
 * Bottom-up DP
 * 
 * 时间：`O(N)`, 0ms
 */

class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        const int n = nums.size();
        if (n == 0) return 0;

        vector<int> up(n, 0); // up[i]表示前i个数字中，最后是上升的最大长度
        vector<int> down(n, 0); // down[i]表示前i个数字中，最后是下降的最大长度
        up[0] = down[0] = 1;

        for (int i = 1; i < n; ++i) {
            up[i] = nums[i] > nums[i-1] ? down[i-1] + 1 : up[i-1]; // 看是否上升
            down[i] = nums[i] < nums[i-1] ? up[i-1] + 1 : down[i-1]; // 看是否下降
        }

        return max(up[n-1], down[n-1]);
    }
};

File: problems/198-house-robber/BottomupDP-v2.js
/**
 * 动态规划：dp[i]表示经历前i个房子能获取的最大价值
 */
var rob = function (nums) {
  const n = nums.length;
  // 状态：dp[i]表示经历前i个房子能获取的最大价值
  const dp = Array.from({ length: n + 1 }, () => 0);
  // 迭代
  for (let i = 1; i <= n; ++i) {
    dp[i] = Math.max(
      dp[i - 1], // 不偷i
      nums[i - 1] + (i - 2 >= 0 ? dp[i - 2] : 0) // 偷i
    );
  }
  // 目标
  return dp[n];
};